pkg/blockchain/noncemanager/redis/manager.go

```diff
@@ -0,0 +1,276 @@
+		
+		return {nonce, cleanupCount}
+	`
+)
+
+// RedisBackedNonceManager implements NonceManager using Redis for persistence.
+// It provides distributed nonce allocation with configurable concurrency limits.
+type RedisBackedNonceManager struct {
+	client    redis.UniversalClient
+	logger    *zap.Logger
+	limiter   *noncemanager.OpenConnectionsLimiter
+	keyPrefix string
+}
+
+// NewRedisBackedNonceManager creates a new Redis-backed nonce manager
+func NewRedisBackedNonceManager(
+	client redis.UniversalClient,
+	logger *zap.Logger,
+	keyPrefix string,
+) *RedisBackedNonceManager {
+	if keyPrefix == "" {
+		keyPrefix = "xmtpd:nonces:"
+	}
+	return &RedisBackedNonceManager{
+		client:    client,
+		logger:    logger.Named("RedisBackedNonceManager"),
+		limiter:   noncemanager.NewOpenConnectionsLimiter(noncemanager.BestGuessConcurrency),
+		keyPrefix: keyPrefix,
+	}
+}
+
+// availableKey returns the Redis key for the available nonces sorted set
+func (r *RedisBackedNonceManager) availableKey() string {
+	return r.keyPrefix + "available"
+}
+
+// sequenceKey returns the Redis key for storing the highest generated nonce
+func (r *RedisBackedNonceManager) sequenceKey() string {
+	return r.keyPrefix + "sequence"
+}
+
+// reservedKey returns the Redis key for the reserved nonces sorted set
+func (r *RedisBackedNonceManager) reservedKey() string {
+	return r.keyPrefix + "reserved"
+}
+
+// GetNonce atomically reserves the next available nonce from Redis.
+// It moves nonces from available set to reserved set to prevent concurrent allocation,
+// similar to how SQL uses SELECT FOR UPDATE SKIP LOCKED.
+func (r *RedisBackedNonceManager) GetNonce(
+	ctx context.Context,
+) (*noncemanager.NonceContext, error) {
+	r.limiter.WG.Add(1)
+
+	// Block until there is an available slot in the blockchain rate limiter
+	select {
+	case r.limiter.Semaphore <- struct{}{}:
+	case <-ctx.Done():
+		r.limiter.WG.Done()
+		return nil, ctx.Err()
+	}
+
+	// Clean up stale reservations and get next available nonce in a single Redis call
+	nonce, err := r.cleanupAndReserveNonce(ctx)
+	if err != nil {
+		r.releaseLimiter()
+		return nil, err
+	}
+
+	metrics.EmitPayerCurrentNonce(float64(nonce))
+
+	return r.createNonceContext(nonce), nil
+}
+
+// Replenish ensures a sufficient number of nonces are available starting from the given nonce.
+// It generates up to 10,000 nonces in a single batch operation using ZADD.
+func (r *RedisBackedNonceManager) Replenish(ctx context.Context, nonce big.Int) error {
+	startNonce := nonce.Int64()
+
+	// Prepare the nonces to add
+	members := make([]redis.Z, BatchSize)
+	for i := int64(0); i < BatchSize; i++ {
+		nonceVal := startNonce + i
+		members[i] = redis.Z{
+			Score:  float64(nonceVal),
+			Member: nonceVal,
+		}
+	}
+
+	// Use a pipeline for efficiency
+	pipe := r.client.Pipeline()
+	pipe.ZAdd(ctx, r.availableKey(), members...)
+	pipe.Set(ctx, r.sequenceKey(), startNonce+BatchSize-1, 0)
```

Comment:
Q: Can you help me understand why we need the sequenceKey? It seems to be like a high watermark, but we never read from it anywhere.
I'd expect something like checking the last sequence key when starting the service, so we replenish from it.

Reply:
Great catch. I was using it in an earlier implementation and didn't clean up